### PR TITLE
Enable auto_vacuum

### PR DESCRIPTION
**Pull request by achocron**
Enables auto vacuum. Tested and working. However, the vacuum (along with all other db set up) is done on the main thread as of now. Leaving the fix for that to another PR.

### CODE REVIEW

**Comment by clpolet** [comment]
@achocron Have you done perf testing to make sure the performances are still ok?

File: AppCenter/AppCenter/Internals/Storage/MSDBStorage.m [hunk 1]
```diff
@@ -112,6 +113,17 @@ + (void)setVersion:(NSUInteger)version inOpenedDatabase:(void *)db {
   [MSDBStorage executeNonSelectionQuery:query inOpenedDatabase:db];
 }
 
++ (void)enableAutoVacuumInOpenedDatabase:(void *)db {
+  NSArray<NSArray *> *result = [MSDBStorage executeSelectionQuery:@"PRAGMA auto_vacuum" inOpenedDatabase:db];
+  int vacuumMode = [(NSNumber *)result[0][0] intValue];
+  BOOL autoVacuumDisabled = vacuumMode == 0;
+  [MSDBStorage executeNonSelectionQuery:@"PRAGMA auto_vacuum = FULL;" inOpenedDatabase:db];
+  if (autoVacuumDisabled) {
+    MSLogDebug([MSAppCenter logTag], @"Vacuuming database to enable auto_vacuum");
```
Comment (dhei) [inline AppCenter/AppCenter/Internals/Storage/MSDBStorage.m]: I suggest write a comment before line 122 to explain why `auto_vacuum` works this way.

There are two scenarios covered: (1) if auto vacuum is enabled, use `PRAGMA auto_vacuum = FULL` ; (2) if auto vacuum is disabled, first use `PRAGMA auto_vacuum = FULL` to set the state, then invoke the VACUUM command to reorganize the entire database file. 

See [auto_vacuum documentation](https://www.sqlite.org/pragma.html#pragma_auto_vacuum) 
```The database connection can be changed between full and incremental autovacuum mode at any time. However, changing from "none" to "full" or "incremental" can only occur when the database is new (no tables have yet been created) or by running the VACUUM command. To change auto-vacuum modes, first use the auto_vacuum pragma to set the new desired mode, then invoke the VACUUM command to reorganize the entire database file.```

File: AppCenter/AppCenter/Internals/Storage/MSDBStorage.m [hunk 2]
```diff
@@ -112,6 +113,17 @@ + (void)setVersion:(NSUInteger)version inOpenedDatabase:(void *)db {
   [MSDBStorage executeNonSelectionQuery:query inOpenedDatabase:db];
 }
 
++ (void)enableAutoVacuumInOpenedDatabase:(void *)db {
+  NSArray<NSArray *> *result = [MSDBStorage executeSelectionQuery:@"PRAGMA auto_vacuum" inOpenedDatabase:db];
+  int vacuumMode = [(NSNumber *)result[0][0] intValue];
+  BOOL autoVacuumDisabled = vacuumMode == 0;
+  [MSDBStorage executeNonSelectionQuery:@"PRAGMA auto_vacuum = FULL;" inOpenedDatabase:db];
```
Comment (dhei) [inline AppCenter/AppCenter/Internals/Storage/MSDBStorage.m]: nitpick - semi-colon at the end of the SQL statement is not needed

**Comment by jaelim-ms** [comment]
Is this something we really need? I carefully read SQLite doc and it says that `auto_vacuum != VACUUM command`. `auto_vacuum` only moves freelist pages to the end of a database file, `it does not defragment the database nor repack individual database pages the way that the VACUUM command does`.
We need to check `auto_vacuum` and see it is really what we want or not.

**Comment by achocron** [comment]
@jaelim-ms yes i think this is important. Based on testing, the benefit is that the database file gets smaller when logs are deleted. I would say that because we are working in a mobile environment, it's fine to save, say, 100 mb of logs temporarily, but it's unacceptable to take 100 mb of storage for a mostly empty database. With auto vacuum, the database file gets smaller once the logs are deleted.

**Comment by achocron** [comment]
@jaelim-ms tests added

File: AppCenter/AppCenterTests/MSDBStorageTests.m [hunk 3]
```diff
@@ -419,6 +419,47 @@ - (void)testDefaultDatabaseSize {
...
+  // If
+
+  // Reset database and ensure that auto_vacuum is enabled.
+  [self.storageTestUtil deleteDatabase];
+  sqlite3 *db = [self.storageTestUtil openDatabase];
+  sqlite3_exec(db, "PRAGMA auto_vacuum = FULL; VACUUM", NULL, NULL, NULL);
+  sqlite3_close(db);
+  id dbStorageMock = OCMClassMock([MSDBStorage class]);
+
+  // Then
```
Comment (dhei) [inline AppCenter/AppCenterTests/MSDBStorageTests.m]: `Then` part is before `When` part

File: AppCenter/AppCenter/Internals/Storage/MSDBStorage.m [hunk 4]
```diff
@@ -110,6 +111,22 @@ + (void)setVersion:(NSUInteger)version inOpenedDatabase:(void *)db {
   [MSDBStorage executeNonSelectionQuery:query inOpenedDatabase:db];
 }
 
++ (void)enableAutoVacuumInOpenedDatabase:(void *)db {
+  NSArray<NSArray *> *result = [MSDBStorage executeSelectionQuery:@"PRAGMA auto_vacuum" inOpenedDatabase:db];
+  int vacuumMode = [(NSNumber *)result[0][0] intValue];
+  BOOL autoVacuumDisabled = vacuumMode != 1;
+
+  /*
+   * If `auto_vacuum` is disabled, change it to `FULL` and then manually `VACUUM` the database. Per the SQLite docs, changing the state of
```
Comment (ElektrojungeMS) [inline AppCenter/AppCenter/Internals/Storage/MSDBStorage.m]: Can we link to the docs that are mentioned?

Comment (dhei) [reply AppCenter/AppCenter/Internals/Storage/MSDBStorage.m]: +1

File: AppCenter/AppCenterTests/MSDBStorageTests.m (hunk 3 above)
Comment (achocron) [reply AppCenter/AppCenterTests/MSDBStorageTests.m]: That's because we need to call `OCMReject` (the "then" part) before the "When" part

**Pull request merged**